packages/bazel-bot/docker-image/generate-googleapis-gen.sh

```diff
@@ -0,0 +1,107 @@
+#!/bin/bash
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Required environment variables:
+# GOOGLEAPIS: path to clone of https://github.com/googleapis/googleapis with
+#   master branch checked out.
+# GOOGLEAPIS_GEN: path to clone of https://github.com/googleapis/googleapis-gen
+#   with master branch checked out.
+#   Git credentials must have been installed so that a git push to $GOOGLEAPIS_GEN
+#   will succeed.
+# BAZEL_REMOTE_CACHE: https path to the bazel remote cache.  Example:
+#   https://storage.googleapis.com/surferjeff-test2-bazel-cache 
+
+# Optional environment variables used for testing.
+# BUILD_TARGETS: Build targets to rebuild.  Example: 
+# //google/cloud/vision/v1:vision-v1-nodejs.tar.gz
+
+# Fail immediately.
+set -e
+
+# Pull both repos to make sure we're up to date.
+git -C "$GOOGLEAPIS" pull
+git -C "$GOOGLEAPIS_GEN" pull
+
+# Collect the history of googleapis.
+shas=$(git -C "$GOOGLEAPIS" log --format=%H)
+
+# Collect shas from googleapis for which we haven't yet generated code in googleapis-gen.
+git -C "$GOOGLEAPIS_GEN" tag > tags.txt
+ungenerated_shas=()
+for sha in $shas; do
+    if grep $sha tags.txt; then
+        # Found a sha we already generated.
+        break
+    else
+        ungenerated_shas+=($sha)
+    fi
+done
+
+# Iterate over the ungenerated_shas from oldest to newest.
+for (( idx=${#ungenerated_shas[@]}-1 ; idx>=0 ; idx-- )) ; do
+    sha="${ungenerated_shas[idx]}"
+
+    # Rebuild at the sha.
+    git -C "$GOOGLEAPIS" checkout "$sha"
+    # Choose build targets.
+    if [[ -z "$BUILD_TARGETS" ]] ; then
+        targets=$(cd "$GOOGLEAPIS" && bazel query 'filter(".*\.tar\.gz$", kind("generated file", //...:*))')
+    else
+        targets="$BUILD_TARGETS"
+    fi
+    # Clean out all the source packages from the previous build.
+    rm -f $(find -L "$GOOGLEAPIS/bazel-bin" -name "*.tar.gz")
+    # Some api always fails to build.
+    # TODO: file a bug when something fails to build.
+    set +e
+    # Invoke bazel build.
+    (cd "$GOOGLEAPIS" && bazel build -k \
+          --remote_cache=$BAZEL_REMOTE_CACHE \
+          --google_default_credentials $targets)
+    
+    # Clear out the existing contents of googleapis-gen.
+    rm -rf "$GOOGLEAPIS_GEN/external" "$GOOGLEAPIS_GEN/google" "$GOOGLEAPIS_GEN/grafeas"
```

Comment:
Do we have  plan to serialize builds, so that we don't end up with potential race conditions on this step?

Reply:
The current algorithm is designed to work correctly with race conditions.  If two builds trigger at the same time, one will win and the other will lose, and the repo will be left in a good state.